### DISCUSSION TITLE
VueFire on nuxt3 does not use runtime environment variables

### DISCUSSION THREAD

**Discussion by SamGeens** [vuejs/vuefire#1363] [Questions and Help] [answered]
### Reproduction

Run vuefire with an at runtime api key does not work

### Steps to reproduce the bug

1. Setup the config in nuxt.config.ts as follows:
```
  modules: [
    'nuxt-vuefire',
  ],

  vuefire: {
    auth: true,
    config: {
      apiKey: process.env.NUXT_PUBLIC_VUEFIRE_API_KEY,
      authDomain: process.env.NUXT_PUBLIC_VUEFIRE_AUTH_DOMAIN,
      projectId: process.env.NUXT_PUBLIC_VUEFIRE_PROJECT_ID,
    },
  },
```
2. Run the app with the environment variables

### Expected behavior

I expect vuefire to use the NUXT_PUBLIC_VUEFIRE_API_KEY environment variable at runtime.

### Actual behavior

The apps shows an empty screen. Which is the exact same behaviour as running the app without api key. Which suggests that the key can't be found at runtime.

**Comment by posva** [comment]
Have you check in that same file the env variables are present? You don’t need to hide those keys btw, they are put on the client code

**Reply by SamGeens** [reply]
@posva The env variables are present. Despite the fact that those keys are put in client code I want to set those keys with env vars at runtime. So I can use it on different environments. 

But it seems the keys can only be set at build time and are not overwritten on runtime.

EDIT: A test repo can be found [here](https://github.com/SamGeens/nuxt3-vuefire-runtimekeys)

**Reply by posva** [reply, marked as answer]
I never saw this, sorry! You need to define them at build time as some of these are inline in the client bundle. I think the best way is to have multiple `.env` files [like suggested in Nuxt docs](https://nuxt.com/docs/guide/directory-structure/env#env-file)

**Reply by SamGeens** [reply]
I indeed now include them at build time. The only downside is that all my other env variables are at runtime so all given in the environment the application runs in but I have to specify the firebase env's during build. 

So while it currently works I was curious if there was a way to use them as runtime vars or that this might be a feature which will be supported in the future.

**Reply by posva** [reply]
It might be possible with public runtime config https://nuxt.com/docs/guide/going-further/runtime-config#example. I will try it out

**Reply by SamGeens** [reply]
That would be great! If you need a test project, mine is still available [here](https://github.com/SamGeens/nuxt3-vuefire-runtimekeys).

**Reply by SamGeens** [reply]
@posva Have you been able to try it like you said?

**Reply by posva** [reply]
The whole config is now public runtime config so it's easy to access. Types still expect one to set the `config` in nuxt.config.ts thought. It might be possible to allow the user to set the config only through env variables but I rather not introduce too many ways of doing the same thing, it just becomes confusing for users. But, I think it's just a Nuxt feature: you can set the runtime env variables, and [they should be applied](https://nuxt.com/docs/migration/runtime-config#runtime-config). It might be a bit verbose through `NUXT_PUBLIC_VUEFIRE_CONFIG_API_KEY` and all the other keys

**Reply by davidstackio** [reply]
@posva It'd be great if this was documented for Nuxt users. I understand showing many ways of doing the same thing can be confusing for folks, but for deploying the same Nuxt app to multiple projects this is incredibly helpful.

I think it makes the most sense to be documented here for Nuxt users (it's also the first place I looked):
https://vuefire.vuejs.org/nuxt/environment-variables.html

For anyone else who comes across this:

### nuxtConfig.ts
(you need to put empty values otherwise you get an error from Vuefire)
```ts
vuefire: {
    config: {
      apiKey: "",
      authDomain: "",
      projectId: "",
      storageBucket: "",
      messagingSenderId: "",
      appId: "",
      measurementId: "",
    },
  },
```

### .env
```
NUXT_PUBLIC_VUEFIRE_CONFIG_API_KEY=
NUXT_PUBLIC_VUEFIRE_CONFIG_AUTH_DOMAIN=
NUXT_PUBLIC_VUEFIRE_CONFIG_PROJECT_ID=
NUXT_PUBLIC_VUEFIRE_CONFIG_STORAGE_BUCKET=
NUXT_PUBLIC_VUEFIRE_CONFIG_MESSAGE_SENDER_ID=
NUXT_PUBLIC_VUEFIRE_CONFIG_APP_ID=
NUXT_PUBLIC_VUEFIRE_CONFIG_MEASUREMENT_ID=
```

**Reply by posva** [reply]
I think it's worth adding that the config is treated as public config and that any env variable [defined using Nuxt convention](https://nuxt.com/docs/guide/going-further/runtime-config#environment-variables) will automatically be picked up by Nuxt. Just in one sentence while referencing nuxt docs. PR Welcome!

**Reply by davidstackio** [reply]
@posva https://github.com/vuejs/vuefire/pull/1562
